docs/designs/anonlink-upload-data-plan.md

```diff
@@ -0,0 +1,289 @@
+# Add data pull feature to Anonlink Entity Service
+
+Author: Brian Thorne
+Status: Draft
+Created: 2020-03-31
+Updated: 2020-04-06
+
+## Abstract/Purpose
+
+In early 2020 blocking has been added to increase the capacity of the Anonlink system, however the system cannot 
+currently handle users uploading large encoded datasets in a reliable way. Investing in scaling Anonlink to a system
+that can deal with tens of millions of records vastly increases the chance of adoption and the general usefulness of 
+the system. 
+
+When a client tries to upload tens of millions of records today they must have a fast and reliable internet connection
+and complete in one REST call the entire data upload operation. These uploads can easily be multiple GiB of data.
+
+On the server side we have had to increase default timeouts to >60s to handle the upload of single digit millions
+of encoded records. The idea of the pull feature is to allow the client to point to a location where the data can be
+pulled by the server. Modifying the client and API to upload in chunks - making many smaller upload requests - was also
+considered. This alternative was rejected as we estimate it would take more effort to get to the same reliability. For
+example the chunking, compression, retrying, and data integrity would all have to be custom handled.
+
+With the Anonlink system we aim to provide everything required to take users from clean records through to an accurate
+concordance table. We rejected the idea of requiring a data provider to setup their own object store or otherwise
+hosting their encoded data online; instead we will modify the client to be able to upload to an object store.
+
+Not including the initial design, implementation is expected to be approximately 2 sprints of work for 2 software 
+engineers. The high level objective is that Anonlink be usable to link between datasets that are roughly the size of 
+the Australian population. 
+
+The aim of the rest of this document is to specify the design of the "pull data" feature.
+
+The sections are:
+
+- Requirements
+- High level system design
+- Specifications for implementation (including the REST API).
+- decisions and alternatives considered. What would be the "best"/"easiest" approach.
+
+
+## Requirements
+
+- A method for requesting temporary & restricted object store credentials shall be added to the Anonlink Entity Service 
+  API. The temporary credentials will authorize the holder to upload data, the credentials **must not** provide access 
+  to any other data providers encodings, or other working data stored in the object store.
```

Comment:
s/providers/providers' (add an apostrophe at the end of this plural to indicate possession -- the data providers have encodings)